File: course/05-byo-image/byo-image.py
"""Lesson 5 (Python, yakyak-sdk): bring your own pre-rendered still, animate it with
Kling AI video, and voice the scene with the cinematic "Narrator" from Lesson 4.

    create-scene(generate:false) -> upload-scene-image -> regen subtitle (voice-over)
    -> rerun from:movie (Kling) -> rerun from:burn (subtitles) -> render.

Run:  pip install -r requirements.txt && python 05-byo-image/byo-image.py  (from course/, after Lesson 1)
"""
import os
import time

import requests
from yakyak_sdk import (ApiClient, Configuration, CreateCampaignDto,
                        CreateSceneDto, DataApi, ExportRenderDto, RerunSceneDto,
                        SaveMovieCustomCastDto, SetCastDto, StartCampaignDto,
                        UpdateCampaignSettingsDto, WorkflowApi)

ROOT = os.path.join(os.path.dirname(__file__), "..")
env = {}
for line in open(os.path.join(ROOT, ".env")):
    line = line.strip()
    if line and not line.startswith("#") and "=" in line:
        k, v = line.split("=", 1)
        env[k.strip()] = v.strip()
BASE, TOKEN, USER = env["YAKYAK_API_BASE"], env["YAKYAK_TOKEN"], env["YAKYAK_USER_ID"]
client = ApiClient(Configuration(host=BASE, access_token=TOKEN))
wf, data = WorkflowApi(client), DataApi(client)

IMAGE = os.path.join(ROOT, "assets/scenes/asian-fruit-lady.jpeg")  # <- your own pre-rendered still
DIALOGUE = "Once upon a time there was a lady who thought her fruits were alive"


def D(x):
    if hasattr(x, "to_dict"):
        return x.to_dict()
    if isinstance(x, list):
        return [D(i) for i in x]
    return x


def upload_scene_image(path, scene_id):
    for _ in range(5):  # the upload occasionally returns an empty body — retry
        try:
            r = requests.post(f"{BASE}/workflow/upload-scene-image",
                headers={"Authorization": f"Bearer {TOKEN}"},
                files={"file": (os.path.basename(path), open(path, "rb"), "image/jpeg")},
                data={"sceneId": scene_id})
            if r.ok and r.json().get("imageUrl"):
                return r.json()["imageUrl"]
        except Exception:
            pass
        time.sleep(2)
    raise RuntimeError(f"scene image upload failed for {path}")


def scene_status(movie_id, scene_id, key):
    m = D(wf.workflow_controller_get_movie(movie_id))
    m = m.get("movie", m)
    sc = next((s for s in (m.get("scene") or m.get("scenes") or []) if s["id"] == scene_id), {})
    return (sc.get(key) or {}).get("status", "waiting")


def wait_asset(movie_id, scene_id, key, label):
    for _ in range(120):
        s = scene_status(movie_id, scene_id, key)
        print(f"  {label}: {s}")
        if s == "completed":
            return
        if s == "failed":
            raise RuntimeError(f"{label} failed")
        time.sleep(5)
    raise TimeoutError(label)


# ---- Setup: a campaign + a Narrator with the cinematic "Cinema" voice (recap of Lesson 4) ----
styles = D(data.data_controller_get_styles())["styles"]
style_id = next((s["id"] for s in styles if "Cartoon" in s["label"]), styles[0]["id"])
campaign_id = D(wf.workflow_controller_create_campaign(CreateCampaignDto.from_dict({
    "userId": USER, "prompt": "A whimsical fruit lady who is convinced her fruits are alive",
    "styleId": style_id, "aspectRatio": "1:1", "animationType": "kenburns", "mode": "pro",
})))["campaignId"]
movie_id = D(wf.workflow_controller_start_campaign(StartCampaignDto.from_dict({"campaignId": campaign_id})))["movieId"]
print("campaign:", campaign_id, "\nmovie:", movie_id)

voices = D(data.data_controller_get_voices())["voices"]
cinema = next((v["voiceId"] for v in voices if v.get("voiceName") == "Cinema"), "Caw0sfpaJco97FKdXypJ")
wf.workflow_controller_save_movie_custom_cast(SaveMovieCustomCastDto.from_dict({"movieId": movie_id, "characters": [
    {"name": "Narrator", "role": "Supporting Character", "description": "A dramatic voice that narrates the story", "sortOrder": 0},
]}))
narrator_id = {c["name"]: c["id"] for c in D(wf.workflow_controller_get_cast(movie_id))["cast"]}["Narrator"]
wf.workflow_controller_set_cast(SetCastDto.from_dict({"movieId": movie_id, "cast": [
    {"id": narrator_id, "name": "Narrator", "role": "Supporting Character", "voiceId": cinema, "fontFamily": "Bangers", "color": "#00abad"},
]}))
print("Narrator ready (cinematic voice).")

# ---- 1) Turn on AI animation (Kling) for this campaign 💸 ----
wf.workflow_controller_update_campaign_settings(UpdateCampaignSettingsDto.from_dict({
    "campaignId": campaign_id, "aspectRatio": "1:1", "animationType": "kling"}))
print("Animation set to Kling (AI video).")

# ---- 2) Create a scene WITHOUT generating art (generate:false) — we bring our own ----
scene_id = D(wf.workflow_controller_create_scene(CreateSceneDto.from_dict({
    "movieId": movie_id, "sceneNumber": 1, "title": "The Fruit Lady", "story": "The fruit lady with her fruits",
    "dialogue": DIALOGUE, "leadCast": "Narrator", "generate": False,
})))["id"]
print("scene:", scene_id)

# ---- 3) Upload your own pre-rendered still as the scene image ✅ ----
img_url = upload_scene_image(IMAGE, scene_id)
print("uploaded your image:", img_url)

# ---- 4) Voice the scene: generate the Narrator's voice-over + captions ✅ ----
wf.workflow_controller_regen_scene_asset(request_body={"sceneId": scene_id, "asset": "subtitle", "from": "subtitle"})
print("Narrating in the cinematic voice…")
wait_asset(movie_id, scene_id, "sceneSubtitleMovie", "voice-over")

# ---- 5) Animate your still with Kling AI video 💸 ----
wf.workflow_controller_rerun_scene(RerunSceneDto.from_dict({"sceneId": scene_id, "from": "movie"}))
print("Kling is animating your image (this can take a few minutes)…")
wait_asset(movie_id, scene_id, "sceneMovie", "kling")

# ---- 6) Burn the subtitles into the animated clip ✅ ----
wf.workflow_controller_rerun_scene(RerunSceneDto.from_dict({"sceneId": scene_id, "from": "burn"}))
print("Burning subtitles…")
wait_asset(movie_id, scene_id, "sceneBurnSubtitle", "subtitles")

# ---- 7) Render → shareable link ----
wf.workflow_controller_export_render(ExportRenderDto.from_dict({"movieId": movie_id, "force": True}))
print("rendering…")
for _ in range(60):
    execs = {e["type"]: e["status"] for e in D(wf.workflow_controller_get_movie_progress(movie_id)).get("executions", [])}
    s = execs.get("movieConcat", "waiting")
    print("  render:", s)
    if s == "completed":
        break
    if s == "failed":
        raise SystemExit("render failed")
    time.sleep(5)
link = ""
for _ in range(36):
    m = D(wf.workflow_controller_get_movie(movie_id))
    m = m.get("movie", m)
    link = m.get("finalMovieUrl") or m.get("soundtrackedMovieUrl") or m.get("concatMovieUrl") or ""
    if link:
        break
    time.sleep(5)
print("🎬 Your movie:", link)

File: course/05-byo-image/byo-image.js
// Lesson 5 (JavaScript, yakyak-sdk): bring your own pre-rendered still, animate it
// with Kling AI video, and voice the scene with the cinematic "Narrator" from Lesson 4.
//   create-scene(generate:false) → upload-scene-image → regen subtitle (voice-over)
//   → rerun from:movie (Kling) → rerun from:burn (subtitles) → render.
// Run:  npm install && node 05-byo-image/byo-image.js   (from course/, after Lesson 1)
import { YakYakClient } from "yakyak-sdk";
import { readFileSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { dirname, join, basename } from "node:path";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");
const env = Object.fromEntries(
  readFileSync(join(ROOT, ".env"), "utf8").split("\n")
    .filter((l) => l.includes("=") && !l.trimStart().startsWith("#"))
    .map((l) => { const i = l.indexOf("="); return [l.slice(0, i).trim(), l.slice(i + 1).trim()]; })
);
const { YAKYAK_API_BASE: base, YAKYAK_TOKEN: token, YAKYAK_USER_ID: userId } = env;
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
const yak = new YakYakClient({ baseUrl: base, token });

const IMAGE = join(ROOT, "assets/scenes/asian-fruit-lady.jpeg");   // <- your own pre-rendered still
const DIALOGUE = "Once upon a time there was a lady who thought her fruits were alive";

// Upload a scene image straight over HTTP (multipart) — same trick Lesson 3/4 use for portraits.
async function uploadSceneImage(path, sceneId) {
  const bytes = await readFile(path);
  for (let i = 0; i < 5; i++) {
    try {
      const fd = new FormData();
      fd.append("file", new Blob([bytes], { type: "image/jpeg" }), basename(path));
      fd.append("sceneId", sceneId);
      const res = await fetch(base + "/workflow/upload-scene-image",
        { method: "POST", headers: { Authorization: `Bearer ${token}` }, body: fd });
      const url = res.ok ? (await res.json().catch(() => ({}))).imageUrl : null;
      if (url) return url;
    } catch { /* retry */ }
    await sleep(2000);
  }
  throw new Error(`scene image upload failed for ${path}`);
}
// Poll one scene asset's status straight off get-movie.
async function sceneStatus(movieId, sceneId, key) {
  const m = await yak.workflow.getMovie({ movieId });
  const mv = m.movie ?? m;
  const sc = (mv.scene ?? mv.scenes ?? []).find((s) => s.id === sceneId) ?? {};
  return (sc[key] ?? {}).status ?? "waiting";
}
async function waitAsset(movieId, sceneId, key, label) {
  for (let i = 0; i < 120; i++) {
    const s = await sceneStatus(movieId, sceneId, key);
    console.log(`  ${label}: ${s}`);
    if (s === "completed") return;
    if (s === "failed") throw new Error(`${label} failed`);
    await sleep(5000);
  }
  throw new Error(`${label} timed out`);
}

// ---- Setup: a campaign + a Narrator with the cinematic "Cinema" voice (recap of Lesson 4) ----
const styles = (await yak.data.getStyles()).styles;
const styleId = (styles.find((s) => s.label.includes("Cartoon")) ?? styles[0]).id;
const { campaignId } = await yak.workflow.createCampaign({ createCampaignDto: {
  userId, prompt: "A whimsical fruit lady who is convinced her fruits are alive",
  styleId, aspectRatio: "1:1", animationType: "kenburns", mode: "pro",
} });
const { movieId } = await yak.workflow.startCampaign({ startCampaignDto: { campaignId } });
console.log("campaign:", campaignId, "\nmovie:", movieId);

const voices = (await yak.data.getVoices()).voices;
const cinema = (voices.find((v) => v.voiceName === "Cinema") || { voiceId: "Caw0sfpaJco97FKdXypJ" }).voiceId;
await yak.workflow.saveMovieCustomCast({ saveMovieCustomCastDto: { movieId, characters: [
  { name: "Narrator", role: "Supporting Character", description: "A dramatic voice that narrates the story", sortOrder: 0 },
] } });
const narratorId = (await yak.workflow.getCast({ movieId })).cast.find((c) => c.name === "Narrator").id;
await yak.workflow.setCast({ setCastDto: { movieId, cast: [
  { id: narratorId, name: "Narrator", role: "Supporting Character", voiceId: cinema, fontFamily: "Bangers", color: "#00abad" },
] } });
console.log("Narrator ready (cinematic voice).");

// ---- 1) Turn on AI animation (Kling) for this campaign 💸 ----
await yak.workflow.updateCampaignSettings({ updateCampaignSettingsDto: { campaignId, aspectRatio: "1:1", animationType: "kling" } });
console.log("Animation set to Kling (AI video).");

// ---- 2) Create a scene WITHOUT generating art (generate:false) — we bring our own ----
const scene = await yak.workflow.createScene({ createSceneDto: {
  movieId, sceneNumber: 1, title: "The Fruit Lady", story: "The fruit lady with her fruits",
  dialogue: DIALOGUE, leadCast: "Narrator", generate: false,
} });
const sceneId = scene.id;
console.log("scene:", sceneId);

// ---- 3) Upload your own pre-rendered still as the scene image ✅ ----
const imgUrl = await uploadSceneImage(IMAGE, sceneId);
console.log("uploaded your image:", imgUrl);

// ---- 4) Voice the scene: generate the Narrator's voice-over + captions ✅ ----
await yak.workflow.regenSceneAsset({ requestBody: { sceneId, asset: "subtitle", from: "subtitle" } });
console.log("Narrating in the cinematic voice…");
await waitAsset(movieId, sceneId, "sceneSubtitleMovie", "voice-over");

// ---- 5) Animate your still with Kling AI video 💸 ----
await yak.workflow.rerunScene({ rerunSceneDto: { sceneId, from: "movie" } });
console.log("Kling is animating your image (this can take a few minutes)…");
await waitAsset(movieId, sceneId, "sceneMovie", "kling");

// ---- 6) Burn the subtitles into the animated clip ✅ ----
await yak.workflow.rerunScene({ rerunSceneDto: { sceneId, from: "burn" } });
console.log("Burning subtitles…");
await waitAsset(movieId, sceneId, "sceneBurnSubtitle", "subtitles");

// ---- 7) Render → shareable link ----
await yak.workflow.exportRender({ exportRenderDto: { movieId, force: true } });
console.log("rendering…");
for (let i = 0, s = "waiting"; i < 60 && s !== "completed"; i++) {
  const prog = await yak.workflow.getMovieProgress({ movieId });
  s = prog.executions?.find((e) => e.type === "movieConcat")?.status ?? "waiting";
  console.log("  render:", s);
  if (s === "failed") { console.error("render failed"); process.exit(1); }
  if (s !== "completed") await sleep(5000);
}
let link = "";
for (let i = 0; i < 36 && !link; i++) {
  const mv = (await yak.workflow.getMovie({ movieId }));
  const m = mv.movie ?? mv;
  link = m.finalMovieUrl || m.soundtrackedMovieUrl || m.concatMovieUrl || "";
  if (!link) await sleep(5000);
}
console.log("🎬 Your movie:", link);
